libbeat/outputs/shipper/shipper.go

```diff
@@ -271,53 +252,113 @@ func (s *shipper) String() string {
 	return "shipper"
 }
 
-func (s *shipper) ackLoop(ctx context.Context, ackClient sc.Producer_PersistedIndexClient) error {
-	st := s.observer
+func (s *shipper) startACKLoop() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	s.ackCancel = cancel
 
+	indexClient, err := s.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+		PollingInterval: durationpb.New(s.config.AckPollingInterval),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to connect to the server: %w", err)
+	}
+	indexReply, err := indexClient.Recv()
+	if err != nil {
+		return fmt.Errorf("failed to fetch server information: %w", err)
+	}
+	s.serverID = indexReply.GetUuid()
+
+	s.log.Debugf("connection to %s (%s) established.", s.config.Server, s.serverID)
+
+	s.ackClient = indexClient
+	s.ackBatchChan = make(chan pendingBatch)
+	s.ackIndexChan = make(chan uint64)
+	s.ackWaitGroup.Add(2)
+
+	go func() {
+		s.ackWorker(ctx)
+		s.ackWaitGroup.Done()
+	}()
+
+	go func() {
+		err := s.ackListener(ctx)
+		s.ackWaitGroup.Done()
+		if err != nil {
+			// TODO: This will stall all acknowledgments if it ever happens.
+			// We should propagate that failure to the Publish calls and/or
+			// try to reconnect.
+			// (Note: this case would be much easier if the persisted index RPC
+			// were not a stream.)
+			s.log.Errorf("acknowledgment listener stopped: %s", err)
```

Comment:
Maybe we should store the status of the ackLoop in the `shipper` struct.  That way we can check it before a publish.

Reply:
I handled this by instead calling `s.Close()` which will produce an error the next time Publish is called.